### PR TITLE
make and travis: cross platform compile check

### PR DESCRIPTION
**Pull request by vbatts**
Signed-off-by: Vincent Batts <vbatts@hashbangbash.com>

### CODE REVIEW

**Comment by vbatts** [comment]
hmmm. `make cross` works for me on my laptop, but not sure why it's getting a permission denied in this travis docker environment.

**Comment by rhatdan** [comment]
Looks like it is actually attempting to create a file on disk as opposed to creating it in the test area?

File: Makefile [hunk 1]
```diff
@@ -7,12 +7,18 @@ SKOPEO_BRANCH = master
 SUDO =
 
 ifeq ($(shell uname),Darwin)
-DARWIN_BUILD_TAG = containers_image_ostree_stub
+PLATFORM_BUILD_TAG = containers_image_ostree_stub containers_image_openpgp
```
Comment (mtrmac) [inline Makefile]: No, non-OpenPGP builds are expected to work natively.  Maybe you want to define a separate cross-build flags variable instead?

File: Makefile [hunk 2]
```diff
@@ -7,12 +7,18 @@ SKOPEO_BRANCH = master
...
+endif
+ifeq ($(GOOS),windows)
+PLATFORM_BUILD_TAG = containers_image_ostree_stub containers_image_openpgp
+endif
+BUILDTAGS = btrfs_noversion libdm_no_deferred_remove $(PLATFORM_BUILD_TAG)
 BUILDFLAGS := -tags "$(BUILDTAGS)"
 
-PACKAGES := $(shell go list $(BUILDFLAGS) ./... | grep -v github.com/containers/image/vendor)
+PACKAGES ?= $(shell go list $(BUILDFLAGS) ./... | grep -v github.com/containers/image/vendor)
```
Comment (mtrmac) [inline Makefile]: Why is this necessary?

File: Makefile [hunk 3]
```diff
@@ -7,12 +7,18 @@ SKOPEO_BRANCH = master
...
 ifeq ($(shell uname),Darwin)
-DARWIN_BUILD_TAG = containers_image_ostree_stub
+PLATFORM_BUILD_TAG = containers_image_ostree_stub containers_image_openpgp
 endif
-BUILDTAGS = btrfs_noversion libdm_no_deferred_remove $(DARWIN_BUILD_TAG)
+ifeq ($(GOOS),darwin)
+PLATFORM_BUILD_TAG = containers_image_ostree_stub containers_image_openpgp
+endif
+ifeq ($(GOOS),windows)
+PLATFORM_BUILD_TAG = containers_image_ostree_stub containers_image_openpgp
```
Comment (mtrmac) [inline Makefile]: Yeah, at three copies, a separate cross-build variable seems much cleaner.

File: Makefile [hunk 4]
```diff
@@ -26,7 +32,11 @@ GPGME_ENV = CGO_CFLAGS="$(shell gpgme-config --cflags 2>/dev/null)" CGO_LDFLAGS=
 all: tools test validate .gitvalidation
 
 build: vendor
-	@$(GPGME_ENV) go build $(BUILDFLAGS) $(PACKAGES)
+	$(GPGME_ENV) go build $(BUILDFLAGS) $(PACKAGES)
```
Comment (mtrmac) [inline Makefile]: Is echoing the hard-to-parse list of packages that useful?

File: Makefile [hunk 5]
```diff
@@ -26,7 +32,11 @@ GPGME_ENV = CGO_CFLAGS="$(shell gpgme-config --cflags 2>/dev/null)" CGO_LDFLAGS=
...
+
+cross:
+	GOOS=windows GOPATH=$(shell mktemp -d):$(GOPATH) $(MAKE) build
+	GOOS=darwin GOPATH=$(shell mktemp -d):$(GOPATH) $(MAKE) build
```
Comment (mtrmac) [inline Makefile]: Wouldn’t it be more accurate to run native builds? At least for macOS this seems possible, e.g. skopeo does that.  Maybe not worth the effort for Windows, sure.

File: Makefile (hunk 1 above)
Comment (vbatts) [reply Makefile]: only when compiling on a darwin host, it can link to gpgme. But if you're cross compiling for darwin then it can't do the gpgme linking

Comment (mtrmac) [reply Makefile]: Right, but this is the `shell uname` condition, when we _are_ on a Darwin host.

File: Makefile (hunk 2 above)
Comment (vbatts) [reply Makefile]: oh it was when i was going to set `PACKAGES` earlier in the makefile, but didn't need to as the build tags filtered the packages that were causing issues.

File: Makefile (hunk 3 above)
Comment (vbatts) [reply Makefile]: i'll refactor a bit

File: Makefile (hunk 4 above)
Comment (vbatts) [reply Makefile]: debugging. I'll add it back.

File: Makefile (hunk 5 above)
Comment (vbatts) [reply Makefile]: at least for a smoke test of the compile. We want this in travis.

**Comment by vbatts** [comment]
updated for @mtrmac comments, but the travis step is still failing. I'm guessing it's the `vendor` target dependencies for the cross-platform targets?

File: Makefile (hunk 1 above)
Comment (vbatts) [reply Makefile]: look at how i've revised it

**Comment by mtrmac** [comment]
> I'm guessing it's the vendor target dependencies for the cross-platform targets?

The `tools.timestamp` target, yes. Just run `go build` inside the `cross` target directly, without recursing into `make build`, perhaps?

**Comment by vbatts** [comment]
nope, it needs the vendor dependency. :grimacing:

File: Makefile [hunk 6]
```diff
@@ -6,10 +6,15 @@ SKOPEO_BRANCH = master
 # Set SUDO=sudo to run container integration tests using sudo.
 SUDO =
 
+ifneq ($(GOOS),linux)
```
[the remark this answers is not in the document]
Comment (mtrmac) [reply Makefile]: Maybe optimize for clarity instead of number of lines?
```
DARWIN_NATIVE_TAGS = …ostree…
DARWIN_CROSS_TAGS = …ostree… …openpgp…
WINDOWS_CROSS_TAGS =…

…

ifeq ($(shell uname),Darwin)
   PLATFORM_BUILD_TAGS = $(DARWIN_NATIVE_TAGS)
endif

…

build: vendor build-internal

build-internal:
	@$(GPGME_ENV) go build $(BUILDFLAGS) $(PACKAGES)

cross:
	make build-internal GOOS=windows PLATFORM_BUILD_TAGS="$(WINDOWS_CROSS_TAGS)"
	make build-internal GOOS=darwin PLATFORM_BUILD_TAGS="$(DARWIN_CROSS_TAGS)"
```

(or with `go build` instead of the new `build-internal` target; all completely untested)

[the remark this answers is not in the document]
Comment (vbatts) [reply Makefile]: i'll give that a try

**Comment by vbatts** [comment]
progress. adding the missing deps now to vendor.conf

**Comment by vbatts** [comment]
weird. The PLATFORM_BUILD_TAGS pass through when i `make cross` locally, but they do not for this travis container build. :confused:

**Comment by vbatts** [comment]
so dumb. The travis setting `BUILDTAGS` is overriding what the makefile has, so the `PLATFORM_BUILD_TAGS` don't get included ...

**Comment by vbatts** [comment]
PTAL

File: Makefile [hunk 7]
```diff
@@ -6,10 +6,18 @@ SKOPEO_BRANCH = master
...
 
+# when cross compiling _for_ a Darwin or windows host, then we must use openpgp
+BUILD_TAG_WINDOWS_CROSS = containers_image_ostree_stub containers_image_openpgp
+BUILD_TAG_DARWIN_CROSS = containers_image_ostree_stub containers_image_openpgp
+# when compiling _on_ a Darwin host, then we can link against gpgme
+BUILD_TAG_DARWIN_NATIVE = containers_image_ostree_stub
+
 ifeq ($(shell uname),Darwin)
-DARWIN_BUILD_TAG = containers_image_ostree_stub
+PLATFORM_BUILD_TAG = BUILD_TAG_DARWIN_NATIVE
```
Comment (mtrmac) [inline Makefile]: Shouldn’t this be `$(BUILD_TAG_DAWIN_NATIVE)`?

**Comment by vbatts** [comment]
Damn you're right. I'll have to fix this in a bit

File: Makefile [hunk 8]
```diff
@@ -25,8 +33,14 @@ GPGME_ENV = CGO_CFLAGS="$(shell gpgme-config --cflags 2>/dev/null)" CGO_LDFLAGS=
...
-build: vendor
-	@$(GPGME_ENV) go build $(BUILDFLAGS) $(PACKAGES)
+build: vendor build-internal
+
+build-internal:
+	$(GPGME_ENV) go build $(BUILDFLAGS) $(PACKAGES)
+
+cross: vendor
+	GOOS=windows $(MAKE) build-internal BUILDTAGS="$(BUILDTAGS) $(BUILD_TAG_WINDOWS_CROSS)"
+	GOOS=darwin $(MAKE) build-internal BUILDTAGS="$(BUILDTAGS) $(BUILD_TAG_DARWIN_CROSS)"
```
Comment (mtrmac) [inline Makefile]: (Non-blocking: This is a tiny bit imprecise, basically it works only because with the current set of tags, adding more tags never breaks a build.

Maybe `.travis.yml` should be overriding `PLATFORM_BUILD_TAG` instead of `BUILDTAGS`?

Non-blocking, it’s likely enough that testing this in all the platforms is not really worth the effort, at least right now.)

File: Makefile [hunk 9]
```diff
@@ -6,10 +6,18 @@ SKOPEO_BRANCH = master
 # Set SUDO=sudo to run container integration tests using sudo.
 SUDO =
 
+# when cross compiling _for_ a Darwin or windows host, then we must use openpgp
+BUILD_TAG_WINDOWS_CROSS = containers_image_ostree_stub containers_image_openpgp
+BUILD_TAG_DARWIN_CROSS = containers_image_ostree_stub containers_image_openpgp
+# when compiling _on_ a Darwin host, then we can link against gpgme
+BUILD_TAG_DARWIN_NATIVE = containers_image_ostree_stub
```
Comment (mtrmac) [inline Makefile]: Oh, and `s/TAG/TAGS/` please.

File: Makefile (hunk 8 above)
Comment (vbatts) [reply Makefile]: i tried a couple of permutations like that. I don't like the way the travis job is just passing through to a docker container, but the current way runs.

**Comment by mtrmac** [comment]
👍 Thanks!
